src/Text/Pandoc/Diagrams.hs

```diff
@@ -1,40 +1,60 @@
-import           Diagrams.Size (dims)
-import           Linear (V2(..), zero)
-import           Options.Applicative
-import           System.Directory                   (createDirectory,
-                                                     doesDirectoryExist)
-import           System.FilePath ((<.>), (</>))
+import qualified Diagrams.Builder                as DB
+import           Diagrams.Prelude                (centerXY, pad, (&), (.~))
+import           Diagrams.Size                   (dims)
+import           Linear                          (V2 (..), zero)
+import           System.Directory                (createDirectory,
+                                                  doesDirectoryExist)
+import           System.FilePath                 ((<.>), (</>), pathSeparator)
 import           System.IO
-import           Text.Pandoc.JSON
+import           Text.Pandoc.Builder             hiding (code, codeBlock)
+import           Text.Pandoc.Generic
 
 -- TODO choose output format based on pandoc target
 backendExt :: String
 backendExt = "png"
 
-main :: IO ()
-main = do
-    opts <- execParser withHelp
-    toJSONFilter $ insertDiagrams opts
-
-insertDiagrams :: Opts -> Block -> IO [Block]
-insertDiagrams opts (CodeBlock (ident, classes, attrs) code)
-    | "diagram-haskell" `elem` classes = (++ [bl']) <$> img
-    | "diagram" `elem` classes = img
-  where
-    img = do
+data Opts = Opts {
+    _outDir     :: FilePath,
+    _expression :: String
+    }
+
+data Echo = Above | Below
+     deriving (Read, Show)
+
+-- | Transform a blog post by looking for code blocks with class
+--   @diagrams@, and replacing them with images generated by evaluating the
+--   identifier @diagrams@ and rendering the resulting diagram.  In
+--   addition, blocks with class @diagrams-def@ are collected (and deleted
+--   from the output) and provided as additional definitions that will
+--   be in scope during evaluation of all @diagrams@ blocks.
+renderBlockDiagrams :: FilePath -> Bool -> Pandoc -> IO Pandoc
+renderBlockDiagrams outDir absolutePath p = bottomUpM (insertDiagrams (Opts outDir "example") absolutePath) p
+
+insertDiagrams :: Opts -> Bool -> Block -> IO Block
+insertDiagrams opts absolutePath (CodeBlock (ident, classes, attrs) code) | "diagrams" `elem` classes = do
         d <- compileDiagram opts code
-        return $ case d of
-            Left _err     -> []
-            Right imgName -> [Plain [Image [] (imgName,"")]] -- no alt text, no title
-    bl' = CodeBlock (ident, "haskell":delete "diagram-haskell" classes, attrs) code
-insertDiagrams _ block = return [block]
+        let imgBlock = case d of
+             Left _err     -> error "diagram not compiling"
+             Right imgName -> Plain [Image [] ((if absolutePath then pathSeparator : imgName else imgName),"")]
+        let codeBlock = CodeBlock (ident, "haskell":delete "diagrams" classes, attrs) code
+        let block' = case readEcho attrs of
```

Comment:
I'm a hesitant about depending on key-value attributes, because they're not well supported by formats besides Markdown.  OTOH, `diagrams-pandoc` doesn't currently handle RST anyway, even though it would be easy.  Maybe if / when we try to use `diagrams-pandoc` for `diagrams-doc`, we'll switch to markdown anyway.  @cchalmers thoughts?